{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import libraries\n",
    "\n",
    "# You need to save \"gw3RACE_functions.py\" file in the same directory to import \"gw3RACE_functions\"\n",
    "import gw3RACE_functions as gw\n",
    "import pandas as pd\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read DataFrame as df\n",
    "df = pd.read_csv('output_Spombe/output_short.tab', sep = '\\t',\n",
    "                 names = ['read_ID', 'chr', 'start_R1', 'stop_R1','strand_R1', 'gene_start', 'gene_stop',\n",
    "                         'gene','coord_R2', 'cigar', 'seq_R2'])\n",
    "\n",
    "## Add new column including 3'RNA tails based on CIGAR code\n",
    "df['tail_fromcigar'] = df.apply(lambda kol: gw.take_tail_fromcigar8(kol.strand_R1, kol.cigar,  \n",
    "                                                                    kol.seq_R2),   axis = 1)\n",
    "\n",
    "## Add new column including 3'RNA tails using grep (only unmapped reads R2)\n",
    "df['tail_fromGREP'] = df.apply(lambda kol: gw.grep_tail_edit_onlyfromSeq(kol.seq_R2), axis = 1)\n",
    "\n",
    "## Add new column including 3'RNA tail sequences based on CIGAR or grep\n",
    "df['tail_GreporCigar'] = df.apply(lambda kol: gw.tail_fromGREPorCIGAR(kol.cigar, kol.tail_fromGREP,\n",
    "                                                                      kol.tail_fromcigar),axis = 1)\n",
    "\n",
    "## Add new column with information 'grep' or 'cigar'                                                                                   \n",
    "df['tail_from'] = df.apply(lambda kol: gw.tail_fromGREPorCIGAR_description(kol.cigar,\n",
    "                                                        kol.tail_fromGREP,kol.tail_fromcigar), axis = 1)\n",
    "\n",
    "## Calculate tail length\n",
    "df['tail_len'] = df['tail_GreporCigar'].apply(lambda x: len(x))\n",
    "\n",
    "## Add new column with information about tail type\n",
    "df['tail_type'] = df.apply(lambda kol:   gw.test_tail_cigargrep8(kol.tail_from, kol.strand_R1, \n",
    "                                                                 kol.tail_GreporCigar),axis = 1)\n",
    "  \n",
    "## Add column with coordinate of 3'end of R2 reads\n",
    "df['stop_R2'] =  df.apply(lambda kol: gw.stop_based_on_cigar(kol.cigar, kol.strand_R1 , \n",
    "                                                             kol.coord_R2),axis = 1)\n",
    "\n",
    "## Calculate distance to transcription end site (TES) based \n",
    "df['distance_to_TES'] =  df.apply(lambda kol: gw.distance_to_TES(kol.cigar,kol.strand_R1 ,kol.stop_R2,\n",
    "                                                              kol.gene_start,kol.gene_stop),axis = 1)\n",
    "\n",
    "## Calculate lentth of gene sequence\n",
    "df['gene_len'] = df['gene_stop']-df['gene_start']\n",
    "\n",
    "## Calculate relative distance to TES\n",
    "df['rel_distance_to_TES'] = df['distance_to_TES']/df['gene_len']\n",
    "\n",
    "## Drop some columns to reduce the size of the table\n",
    "df.drop(columns= [ 'chr', 'start_R1', 'stop_R1', 'strand_R1', 'gene_start',\n",
    "       'gene_stop',  'coord_R2', 'cigar', 'seq_R2', 'tail_fromcigar',\n",
    "       'tail_fromGREP', 'stop_R2', 'gene_len'], inplace = True)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Assign RNA type - it will works only for S. pombe genome\n",
    "\n",
    "df['RNA_type'] = df['gene'].apply(lambda x: gw.test_RNA_type(x.lower()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Save DataFrame as csv\n",
    "df.to_csv('output_Spombe/output_detailed_DataFrame.csv', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>read_ID</th>\n",
       "      <th>gene</th>\n",
       "      <th>tail_GreporCigar</th>\n",
       "      <th>tail_from</th>\n",
       "      <th>tail_len</th>\n",
       "      <th>tail_type</th>\n",
       "      <th>distance_to_TES</th>\n",
       "      <th>rel_distance_to_TES</th>\n",
       "      <th>RNA_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A00805:163:HFM2WDRX2:1:1101:1000:1626:ANTGTT</td>\n",
       "      <td>SPCC320.14</td>\n",
       "      <td></td>\n",
       "      <td>cigar</td>\n",
       "      <td>0</td>\n",
       "      <td>no_tail</td>\n",
       "      <td>-482</td>\n",
       "      <td>-0.243557</td>\n",
       "      <td>mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A00805:163:HFM2WDRX2:1:1101:1000:19445:CNCCTG</td>\n",
       "      <td>SPBC947.02</td>\n",
       "      <td></td>\n",
       "      <td>cigar</td>\n",
       "      <td>0</td>\n",
       "      <td>no_tail</td>\n",
       "      <td>-729</td>\n",
       "      <td>-0.218656</td>\n",
       "      <td>mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A00805:163:HFM2WDRX2:1:1101:1000:27085:CTTTAT</td>\n",
       "      <td>SPSNRNA.07</td>\n",
       "      <td></td>\n",
       "      <td>cigar</td>\n",
       "      <td>0</td>\n",
       "      <td>no_tail</td>\n",
       "      <td>-1</td>\n",
       "      <td>-0.003922</td>\n",
       "      <td>snRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A00805:163:HFM2WDRX2:1:1101:10004:28682:TCACAC</td>\n",
       "      <td>SPBC1539.09c</td>\n",
       "      <td>TTTTTTTTTTTTTT</td>\n",
       "      <td>cigar</td>\n",
       "      <td>14</td>\n",
       "      <td>polyA</td>\n",
       "      <td>130</td>\n",
       "      <td>0.057018</td>\n",
       "      <td>mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A00805:163:HFM2WDRX2:1:1101:10004:7075:TCTTCA</td>\n",
       "      <td>SPSNRNA.07</td>\n",
       "      <td></td>\n",
       "      <td>cigar</td>\n",
       "      <td>0</td>\n",
       "      <td>no_tail</td>\n",
       "      <td>-1</td>\n",
       "      <td>-0.003922</td>\n",
       "      <td>snRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          read_ID          gene  \\\n",
       "0    A00805:163:HFM2WDRX2:1:1101:1000:1626:ANTGTT    SPCC320.14   \n",
       "1   A00805:163:HFM2WDRX2:1:1101:1000:19445:CNCCTG    SPBC947.02   \n",
       "2   A00805:163:HFM2WDRX2:1:1101:1000:27085:CTTTAT    SPSNRNA.07   \n",
       "3  A00805:163:HFM2WDRX2:1:1101:10004:28682:TCACAC  SPBC1539.09c   \n",
       "4   A00805:163:HFM2WDRX2:1:1101:10004:7075:TCTTCA    SPSNRNA.07   \n",
       "\n",
       "  tail_GreporCigar tail_from  tail_len tail_type  distance_to_TES  \\\n",
       "0                      cigar         0   no_tail             -482   \n",
       "1                      cigar         0   no_tail             -729   \n",
       "2                      cigar         0   no_tail               -1   \n",
       "3   TTTTTTTTTTTTTT     cigar        14     polyA              130   \n",
       "4                      cigar         0   no_tail               -1   \n",
       "\n",
       "   rel_distance_to_TES RNA_type  \n",
       "0            -0.243557     mRNA  \n",
       "1            -0.218656     mRNA  \n",
       "2            -0.003922    snRNA  \n",
       "3             0.057018     mRNA  \n",
       "4            -0.003922    snRNA  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
